vault/resource_kmip_secret_role.go

```diff
@@ -0,0 +1,287 @@
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Computed:    true,
+				Description: "Grant permission to use the KMIP Locate operation",
+			},
+			"operation_register": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Computed:    true,
+				Description: "Grant permission to use the KMIP Register operation",
+			},
+			"operation_rekey": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Computed:    true,
+				Description: "Grant permission to use the KMIP Rekey operation",
+			},
+			"operation_revoke": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Computed:    true,
+				Description: "Grant permission to use the KMIP Revoke operation",
+			},
+		},
+	}
+}
+
+func kmipSecretRoleCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Get("path").(string)
+	scope := d.Get("scope").(string)
+	role := d.Get("role").(string)
+
+	data := map[string]interface{}{}
+	if v, ok := d.GetOk("tls_client_key_bits"); ok {
+		data["tls_client_key_bits"] = v.(int)
+	}
+
+	if v, ok := d.GetOk("tls_client_key_type"); ok {
+		data["tls_client_key_type"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("tls_client_ttl"); ok {
+		data["tls_client_ttl"] = v.(int)
+	}
+
+	for _, k := range kmipRoleAPIBooleanFields {
+		if v, ok := d.GetOkExists(k); ok {
+			data[k] = v.(bool)
+		}
+	}
+
+	scopePath := path + "/scope/" + scope
+	rolePath := scopePath + "/role/" + role
+	log.Printf("[DEBUG] Updating %q", rolePath)
+	if _, err := client.Logical().Write(rolePath, data); err != nil {
+		return fmt.Errorf("error updating KMIP config %q, err=%w", rolePath, err)
+	}
+
+	d.SetId(path)
+	d.Set("scope", scope)
+	d.Set("role", role)
+
+	return kmipSecretRoleRead(d, meta)
+}
+
+func kmipSecretRoleRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+	scope := d.Get("scope").(string)
+	role := d.Get("role").(string)
+
+	scopePath := path + "/scope/" + scope
+	rolePath := scopePath + "/role/" + role
+	log.Printf("[DEBUG] Reading KMIP role at %q", rolePath)
+	resp, err := client.Logical().Read(rolePath)
+	if err != nil {
+		return fmt.Errorf("error reading KMIP role at %s, err=%w", rolePath, err)
+	}
+	if resp == nil {
+		log.Printf("[WARN] KMIP role not found, removing from state")
+		d.SetId("")
+		// TODO Confirm this behavior
+		d.Set("scope", "")
+		d.Set("role", "")
+		return fmt.Errorf("expected role at %s, no role found", rolePath)
+	}
+
+	d.Set("tls_client_key_bits", resp.Data["tls_client_key_bits"])
+	d.Set("tls_client_key_type", resp.Data["tls_client_key_type"])
+	d.Set("tls_client_ttl", resp.Data["tls_client_ttl"])
+
+	for _, k := range kmipRoleAPIBooleanFields {
+		if err := d.Set(k, resp.Data[k]); err != nil {
+			return fmt.Errorf("error setting state key %q on KMIP config, err=%w", k, err)
+		}
+	}
+
+	return nil
+}
+
+func kmipSecretRoleUpdate(d *schema.ResourceData, meta interface{}) error {
```

Comment:
`kmipSecretRoleUpdate` looks very similar to `kmipSecretRoleCreate`, I wonder if it might be worth using the same function for Create and Update, or perhaps factoring out some of the common bits?

Reply:
Curious if you still think this is the case after the addition of remounts

Reply:
Added a refactor in 6c67a14